Doc/library/asyncio-stream.rst

```diff
@@ -367,13 +366,167 @@ Stream
+      * :attr:`StreamMode.READWRITE` - Connection can send and receive data.
+
+   .. coroutinemethod:: abort()
+
+      Aborts the connection immediately, without waiting for the send buffer to drain.
+
+   .. coroutinemethod:: drain()
+
+      Wait until it is appropriate to resume writing to the stream.
+      Example::
+
+          stream.write(data)
+          await stream.drain()
+
+      This is a flow control method that interacts with the underlying
+      IO write buffer.  When the size of the buffer reaches
+      the high watermark, *drain()* blocks until the size of the
+      buffer is drained down to the low watermark and writing can
+      be resumed.  When there is nothing to wait for, the :meth:`drain`
+      returns immediately.
+
+   .. coroutinemethod:: sendfile(file, offset=0, count=None, *, fallback=True)
+
+      Calls :meth:`Stream.drain()` to write to the connection and uses :meth:`loop.sendfile`
+      to send *file* over *transport*.
+
+   .. coroutinemethod:: start_tls(sslcontext, *, server_hostname=None, \
+                                  ssl_handshake_timeout=None)
+
+      Calls :meth:`Stream.drain()` to write to the connection and uses :meth:`loop.start_tls`
+      to upgrade the existing transport-based connection to TLS.
+
+   .. coroutinemethod:: read(n=-1)
+
+      Read up to *n* bytes.  If *n* is not provided, or set to ``-1``,
+      read until EOF and return all read bytes.
+
+      If EOF was received and the internal buffer is empty,
+      return an empty ``bytes`` object.
+
+   .. coroutinemethod:: readline()
+
+      Read one line, where "line" is a sequence of bytes
+      ending with ``\n``.
+
+      If EOF is received and ``\n`` was not found, the method
+      returns partially read data.
+
+      If EOF is received and the internal buffer is empty,
+      return an empty ``bytes`` object.
+
+   .. coroutinemethod:: readexactly(n)
+
+      Read exactly *n* bytes.
+
+      Raise an :exc:`IncompleteReadError` if EOF is reached before *n*
+      can be read.  Use the :attr:`IncompleteReadError.partial`
+      attribute to get the partially read data.
+
+   .. coroutinemethod:: readuntil(separator=b'\\n')
+
+      Read data from the stream until *separator* is found.
+
+      On success, the data and separator will be removed from the
+      internal buffer (consumed). Returned data will include the
+      separator at the end.
+
+      If the amount of data read exceeds the configured stream limit, a
+      :exc:`LimitOverrunError` exception is raised, and the data
+      is left in the internal buffer and can be read again.
+
+      If EOF is reached before the complete separator is found,
+      an :exc:`IncompleteReadError` exception is raised, and the internal
+      buffer is reset.  The :attr:`IncompleteReadError.partial` attribute
+      may contain a portion of the separator.
+
+   .. method:: write(data)
+
+      The method attempts to write the *data* to the underlying socket immediately.
```

Comment:
I think here and below for methods that can be used both with `await` and without it we should focus on `await` version first.  E.g. `await stream.write(data)` is more idiomatic than `stream.write(data); await stream.drain()`.

The first documentation line can be `Write *data* to the underlying socket; wait until the data is sent.`.

Thinking more, I believe we need to document two functions with the same name:
```
,, comethod:: write(data)

   Write *data* to the underlying socket; wait until the data is sent, e.g.::

      await stream.write(b'data')

,, method:: write(data)

   Long description here that teaches how combite `write()` with `drain()` and mentions that usually people should prefer `await stream,write()` version.
```
I'm curious is it supported by sphinx well or not? What version is a target for `:meth:Stream.write` expression?

Reply:
It is supported. Here is a first pass version of this. The `Stream.write` target is set to the coroutine and the normal write method cannot be hyperlinked. We also do some explanation on the drain method so I have the docs as below : 

```rst
   .. coroutinemethod:: write(data)

      Write *data* to the underlying socket; wait until the data is sent, e.g.::

         await stream.write(data)

   .. method:: write(data)

      The method attempts to write the *data* to the underlying socket immediately.
      If that fails, the data is queued in an internal write buffer until it can be
      sent. :meth:`drain` can be used to flush the underlying buffer once writing is
      available::

         stream.write(data)
         await stream.drain()

      It is recommended to directly await on the `write()` method instead::

         await stream.write(data)
```

![Screen Shot 2019-09-11 at 10 59 30 am](https://user-images.githubusercontent.com/3972343/64687956-5a29bd80-d483-11e9-9d2e-e7524ee4c356.png)

Reply:
Looks nice!